### ISSUE TITLE
Parallelised reporting

### ISSUE THREAD

**Issue by jhmatthews** [sirocco-rt/sirocco#28]
Now the code has been parallelised, we need to agree on how to process diagnostics in parallel mode.
- Do we still want to keep all the diagnostic files for each thread? I suspect so. In which case I propose we keep them in a subfolder with the name diag_root, for organisational purposes. 
- Same for signal files? in which case we need to modify the xsignal routine slightly so it has a files for each thread like the diag files
- If we do this, the directory after running cv.pf would look a little like this

![screen shot 2013-07-16 at 14 04 29](https://f.cloud.github.com/assets/3007249/804757/4cad427e-ee18-11e2-8404-93c74b3b6e9f.png)
- at the moment, when running in parallel mode python will print out a full set of prints for everything...so can we change the Log statement so it only prints to screen if we are in MPI mode? I suspect that might be diffifcult with kpar as it's used as a library and we don't want to call Log with a variable...
- note that this is another reason for us not to use print statements if we can help it

Thoughts, @kslong and @Higginbottom ?

**Comment by kslong** [comment]
I think we need to really think this through as it is quite complicated.  There are lots of "use cases" here that were developed over the years for the single string and I am not sure how to proceed exactly.   I do think the basic idea of put the diagnostic files in a subdirectory is going to be required, though I would probably name this cv.diag instead of diag.cv, just to keep everything iasscocaited with one run together in a directory listing.

I probably favor one subdirectory rather than two.

One good thing if we have a unique subdirectory name, is that we should be able to run different models from the same directory at the same time, as long as the .pf file is unique.  Right now there are some extra diagnostic files that are produced that have fixed names and these can be put in the subdirectory.

We need to figure out whether recomb.save is needed at all.  I'm unsure how many times it is being created and whether it is really saving us time.

sig was originally created so that we could restart python in single string mode.  It was designed for the case where everything was job parallel, or one was on a beowulf cluster and you expected to  have to use restarts.  I'm not quite sure how this works in multi string mode.  My first guess is that this really needs to be a single file, and that it should only be written to by the master process, if that is possible.

(I also think it will be simpler if the .pf file is read by the master process and the results distributed)

Clearly there are going to have to be multiple .diag files, since errors are created during the process of transmitting photons.  But if we just leave it at that, we are going to need some kind of helper routine to sum up what has happened, and whether the program has actually completed in the way that we want.   This could be a separate (python) routine, I guess.

These are just my initial thoughts, but how we do this is going to be important, especially for making python public.  It needs to run cleanly both in atingle and multiple string mode.

Cheers,
Knox

**Comment by jhmatthews** [comment]
RE the sig file- at the moment it writes to it for each thread. In some ways this is useful, but the sig file does it fairly infrequently, so I think just having the master thread write to sig should be fine. To do that, the simplest thing is just to put theses lines in signal.c

``` C
#ifdef MPI_ON
  if (rank_global==0)
    {
#endif

...the code in the function

#ifdef MPI_ON
  }
#endif
```

The only problem I can't really think of a way round at the moment is how to stop python printing out to screen for each thread, since the log statement knows nothing about which thread it is being called on...

Edit: actually, I think that last problem is pretty easy

a function in kpar, which is called near the start of python.c

``` C
int Log_set_mpi_rank(rank)
    int rank;
{
    my_rank=rank;
    return(0);
}
```

which just copies a variable passed from python.c down to log means we can make it so we only print to screen if my_rank=0. This is now implemented- download new version of kpar from ~/Dropbox/Python, or simply paste the above function into the bottom of log.c. main() calls this function near the start of the routine- this is in the dev branch and will be in a release soon.

I can do something similar to collect the error summaries together, but that will also involve communicating between threads.

**Comment by kslong** [comment]
There is a verbosity parameter in the logging routines, which might be used to prevent, printing to the screen, assuming we rigidly avoid printf.

Note that kpar itself does some printing to the screen itself, so that it can operate in interactive mode.   This may need to be handled separately.   This is yet another argument that one gets all the inputs from thread 0 and then we promulgate them to all threads before running the photionization cycles.

Knox

**Comment by jhmatthews** [comment]
Related to this, I just wanted to note that I don't think reporting to stderr (as we now do in trans_phot) is a good idea in general- an example being that on Iridis this gets put in your error file for the batch script and not your output file.

**Comment by kslong** [comment]
Agreed.  I think we need a proposed list of fixes for this general problem that we implement sequentially.  The first two items on the list should be
1.  Eliminate all normal outputs to stderr, either completely or by converting to Log commands
2.  Eliminate all printf statements, in the same manner

Note kpar may be a separate issue

Knox S. Long
Space Telescope Science Institute

**Comment by jhmatthews** [comment]
My initial list, some of which has been started
- Eliminate all normal outputs to stderr, either completely or by converting to Log commands 
  **(I've reverted to old method for the moment in dev branch)**
- Eliminate all printf statements, in the same manner (NSH is doing this.)
- I think we probably want a separate log statement, like Log_parallel, which reports on how the threads are getting on and prints for all threads, as well as to the diag file- the reason I say this is because there are useful statements like 'thread 10 is happy after broadcast' which I think are useful to print to screen. This can then be used for the occasional reason one might want to print to screen for all threads (an example being errors, possibly).
- work out how we are going to deal with error counting and reporting. This is not trivial as I think the error_count arrays are managed by kpar. I think if we have them global in log.h that might work in terms of communicating them and counting up between threads. We will then need an error_gather function to do this, either after each cycle, or just at the end for the error summary, and change the max error counts by dividing by nprocessors?
  **Diag files, python.phot and disk.diag files are now saved in a subdirectory diag_root. the script py_error.py provides an error summary of the threads. It can be found on Dropbox/Python/**
- work out how to deal with the pf file readin. Proper fix is probably reading in by the master thread then sending info to slave threads. This might be a bit fiddly. Alternative would be only print to screen in the 0th thread by modifying rdpar and rdchar slightly, like I did for the log statements.

There are a lot of print statements commented out. Should these all be deleted, without exception? I wonder if some may be useful to retain.

**Comment by jhmatthews** [comment]
@kslong @Higginbottom I've been working on reporting and verbosity a little today. This has involved some editing of both python.c and **kpar (new version in ~/Dropbox/Python/reporting/)**

Basically, in parallel mode, we now work like this. when reading the pf file, only the master thread is printed to screen. 

**run with -v 0: Nothing printed, except stderr like 'could not rm .sig file'**

**run with -v 1: Only parallel statements printed.**
An example printed output for a one ionization cycle run:

```
Thread 3 starting.
Thread 0 starting.
Thread 1 starting.
Thread 2 starting.
This is MPI task number 1 (a total of 4 tasks are running).
This is MPI task number 3 (a total of 4 tasks are running).
This is MPI task number 2 (a total of 4 tasks are running).
This is MPI task number 0 (a total of 4 tasks are running).
Zeroth thread successfully received the normalised estimators. About to broadcast.
Thread 1 happy after broadcast.
Thread 3 happy after broadcast.
Thread 0 happy after broadcast.
Thread 2 happy after broadcast.
Zeroth thread successfully received the integer sum. About to broadcast.
Thread 3 Finalized. All done
Thread 1 Finalized. All done
Thread 0 Finalized. All done
Thread 2 Finalized. All done
```

**run with -v 2: Log statements for main thread are printed to screen.**

This will look much like a normal one processor run- errors and so on are reported to screen. Check ~/Dropbox/Python/reporting/output_v2 for an example output of one ionization cycle with cv_bb_nb.pf.

Not entirely sure if the way I've implemented this is the cleanest solution. It strikes me as a little messy, but there appeared to be messiness associated with most solutions I could think of. Any feedback appreciated! This is at least a reasonable solution for the moment.

Once we've sorted this and Nick has tested the bugfixes we'll make a new release.

**Comment by jhmatthews** [comment]
Right, we are getting there on this. I am fairly happy with most of the above, but we need to decide a way to do the error reporting.

**1) error counting- am I OK to simply change the following in kpar?**

``` C
 int log_print_max=100;           // Mximum number of times a single error will be reported.  Note that 
                // Note that it will still be counted.
int time_to_quit=1000000;
```

with a function that does this:

``` C
int Log_set_mpi_rank(rank)
    int rank;
{
    my_rank=rank;
    rdpar_set_mpi_rank(rank);  //this just communicates the rank to rdpar
        time_to_quit /= n_mpi;
         log_print_max /= n_mpi;
    return(0);
}
```

**2) Which errors do we want printed to screen? Options are:**
- none. they just go to diags and you should check diags
- all errors for all threads (could get messy...)
- my favoured option: none printed to screen, or just the 0th thread printed to screen, and then the user has a script they can run which gives an error summary while the run is in progress, which tells you the number of errors and the threads they occur in

**Comment by kslong** [comment]
This sounds reasonable to me, James.  I think we should try it and see how we like it.
Cheers,
Knox

Knox S. Long
Space Telescope Science Institute

**Comment by jhmatthews** [comment]
Ok, I have done so.

### kpar

I didn't actually change the log_print_max  variable, as I think it's sensible to keep this at 100 for each thread, especially as we only print one thread to screen. the time to quit one is changed in kpar_76b.tar.gz on Dropboz

### scripts

We now have two scripts, found in ~/Dropbox/Python/reporting/

**watchdog.py**
usage: python watchdog.py [-f folder] root
default folder: diag_root

You can run this on an ongoing run. it will look for Errors in the diag files in the folder diag_root or a specified folder. It will then collate these errors and provide an onscreen output. You can see a sample output for it: ~/Dropbox/Python/reporting/watchdog_output.

It prints 
- the total errors across all thread - note that these aren't the actual error counts but just the ones that are logged
- however, it then tells you which ones have exceeded the error count limits and which threads they are in
- it then gives you an ongoing error summary for each thread

**py_error.py**
usage: python py_error.py [-f folder] root
default folder: diag_root

This is the post processing script. It goes through the error summaries for each thread and counts them all. You can see a sample output for it: ~/Dropbox/Python/reporting/py_error_output.

We may decide we want to do more with these scripts, but for the moment they work and are useful!

In terms of install, these should probably go in the py_progs directory or something similar, then get chmod +x'ed and copied to bin in the configuration of Python.

@kslong - note recent changes to kpar, and routines in ~/Dropbox/Python/reporting/

I'm leaving this issue open as we will probably want to make improvements, but it's certainly getting there.

**Comment by jhmatthews** [comment]
note that kpar is now incorporated into python and not used as a library. I'm closing this issue for now as reporting is done much better, although could doubtless be improved. Error logging needs to be dealt with and is addressed in #47.
